### PR TITLE
test(migrations): parametrize rollback tests across postgres/mysql/sqlite

### LINKED ISSUE #4557: test(integration): extend migration rollback test matrix to MySQL and SQLite backends
### Prerequisites

- [x] I have searched existing issues to verify this feature hasn't been requested

### Summary

Extend the migration rollback integration test suite (currently in `tests/integration/tests/migrations/migration_rollback_integration.rs`, 13 cases) to **MySQL** and **SQLite** backends. Today these tests run only against PostgreSQL via TestContainers, leaving DB-specific rollback edge cases unverified for the other supported backends.

### Motivation / Problem

The rollback behavior is **not portable** across the three supported backends due to fundamental DDL semantics:

| Backend | Transactional DDL | Implication for rollback |
|---------|-------------------|--------------------------|
| PostgreSQL | ✅ Full | Atomic rollback of `CREATE TABLE` + `ALTER` works via `BEGIN`/`ROLLBACK` |
| MySQL | ❌ Non-transactional | DDL auto-commits; partial-rollback semantics differ — `reverse_sql` must compensate explicitly |
| SQLite | ⚠️ Partial | `ALTER COLUMN` is limited (table recreation required); rollback path is different |

The current PostgreSQL-only coverage means:

1. **MySQL non-transactional DDL bugs are invisible** — e.g., a failure mid-rollback leaves the schema in an inconsistent state that the test suite cannot detect.
2. **SQLite table-recreation rollback** is only partially covered (`sqlite_table_recreation_integration.rs::test_recreation_transaction_rollback`) and not aligned with the 13-case matrix.
3. **Cross-backend regression risk** — a refactor to `executor.rs` could silently break MySQL while PostgreSQL CI stays green.

### Proposed Solution

Refactor the 13 existing rollback tests into a backend-parameterized form using `rstest` with a backend fixture, so each case runs against PostgreSQL, MySQL, and SQLite in turn:

```rust
#[rstest]
#[case::postgres(BackendKind::Postgres)]
#[case::mysql(BackendKind::Mysql)]
#[case::sqlite(BackendKind::Sqlite)]
#[tokio::test]
async fn test_create_table_rollback(#[case] backend: BackendKind) {
    // Arrange: spin up backend-specific container/in-memory DB via reinhardt-test fixture
    // ...
}
```

Where backend-specific behavior diverges (e.g., MySQL non-transactional DDL), use **separate cases or skip annotations with explicit reasoning** rather than silently passing — the test should either verify the divergent behavior or be marked `#[ignore]` with a comment pointing to a tracking issue.

**Cases to port (all 13):**

1. `test_create_table_rollback`
2. `test_add_column_rollback`
3. `test_alter_column_type_rollback`
4. `test_run_sql_with_reverse_rollback`
5. `test_atomic_multi_operation_rollback` — expected to **diverge** on MySQL (non-atomic)
6. `test_rollback_with_data`
7. `test_index_rollback`
8. `test_rollback_fail_without_reverse_sql`
9. `test_partial_rollback_non_atomic`
10. `test_rollback_fail_with_foreign_key_reference`
11. `test_rollback_empty_migration`
12. `test_rollback_with_dependencies`
13. `test_circular_dependency_rollback`

**TestContainers fixtures needed:**

- `mysql_container` (likely already exists in `reinhardt-test::fixtures` — verify)
- `sqlite_in_memory` or `sqlite_file` fixture (no container needed)

### Alternatives Considered

- **Test only the common subset across backends** — rejected. The DB-specific edge cases (MySQL non-transactional DDL, SQLite ALTER limitations) are exactly where bugs hide and most need coverage.
- **Manual smoke tests per release** — rejected. Not reproducible, not regression-safe, doesn't scale.

### Additional Context

- Existing tests to port: `tests/integration/tests/migrations/migration_rollback_integration.rs` (~1100 lines, 13 cases)
- Related backend-specific test (already exists, do not duplicate): `tests/integration/tests/migrations/sqlite_table_recreation_integration.rs:1285::test_recreation_transaction_rollback`
- Per `CLAUDE.md`: use TestContainers (Docker, not Podman); fixtures live in `reinhardt-test`
- Per `instructions/TESTING_STANDARDS.md` TI-1: use `rstest` + AAA pattern (`// Arrange` / `// Act` / `// Assert`)
- This issue depends on **none** of the other rollback issues — can be worked independently

### PR DESCRIPTION
**Pull request by kent8192**
## Summary

Parametrize the migration rollback integration tests to run against all three supported database backends (Postgres, MySQL, SQLite) using `rstest::rstest` + `#[case]`. All 13 existing test cases now execute on each backend (39 instances total).

## Type of Change

- [x] Test coverage expansion (non-breaking)

## Motivation and Context

Previously the rollback test suite only validated the Postgres path. Backends with non-transactional DDL (MySQL) and backends with limited ALTER COLUMN support (SQLite) had no cross-backend rollback coverage, leaving regressions in the executor invisible to CI. See #4557.

## How Was This Tested?

- `cargo check -p reinhardt-integration-tests --test migrations --all-features` — clean
- `cargo clippy -p reinhardt-integration-tests --test migrations --all-features --no-deps` — file is clippy-clean (no new warnings introduced by this PR; preexisting warnings in other migration test files are out of scope)
- `cargo fmt -- --check tests/integration/tests/migrations/migration_rollback_integration.rs` — clean
- `cargo nextest run -p reinhardt-integration-tests --test migrations migration_rollback_integration --all-features` — **34 / 39 pass; 5 fail**
- TestContainers via Docker (Podman explicitly avoided per `CLAUDE.md`)

### Test Matrix Result

| Test | Postgres | MySQL | SQLite |
|---|---|---|---|
| `test_create_table_rollback` | PASS | PASS | PASS |
| `test_add_column_rollback` | PASS | PASS | PASS |
| `test_alter_column_type_rollback` | PASS | **FAIL** (#4582) | **FAIL** (#4582) |
| `test_run_sql_with_reverse_rollback` | PASS | PASS | PASS |
| `test_atomic_multi_operation_rollback` | PASS | PASS | **FAIL** (#4584) |
| `test_rollback_with_data` | PASS | PASS | PASS |
| `test_index_rollback` | PASS | **FAIL** (#4583) | PASS |
| `test_rollback_fail_without_reverse_sql` | PASS | PASS | PASS |
| `test_partial_rollback_non_atomic` | PASS | PASS | PASS |
| `test_rollback_fail_with_foreign_key_reference` | PASS | PASS | PASS |
| `test_rollback_empty_migration` | PASS | PASS | PASS |
| `test_rollback_with_dependencies` | PASS | **FAIL** (#4585) | PASS |
| `test_circular_dependency_rollback` | PASS | PASS | PASS |

### Discovered Bugs

The 5 failures are **real pre-existing bugs in `reinhardt-db`** that the new cross-backend coverage has exposed. Per `CLAUDE.md` § Test Quality Rules, the assertions must NOT be weakened to make tests pass; the failures are tracked in follow-up issues:

- #4582 — `ALTER COLUMN TYPE` generates Postgres-only syntax on MySQL/SQLite
- #4583 — MySQL `DROP INDEX` produces malformed SQL
- #4584 — SQLite atomic multi-operation rollback hits "database schema is locked"
- #4585 — MySQL FK-dependent rollback times out acquiring migration lock

This is **exactly the value the issue described**:

> "MySQL non-transactional DDL bugs are invisible — e.g., a failure mid-rollback leaves the schema in an inconsistent state that the test suite cannot detect."

This PR exposes those bugs without weakening the assertions. The fixes belong in separate per-bug PRs (per `instructions/ISSUE_HANDLING.md` work-unit principles: 1 PR = 1 crate × 1 fix pattern).

## Design Notes

- New local `setup_backend(DatabaseType)` returns `(BackendHandle, DatabaseConnection)`; `BackendHandle` owns the container so it is torn down on test exit. SQLite uses `sqlite::memory:` (no container needed).
- Backend-aware introspection helpers (`table_exists`, `column_exists`, `index_exists`) dispatch on `DatabaseConnection::database_type()` (uses `information_schema` for Postgres/MySQL, `sqlite_master` / `PRAGMA table_info` for SQLite).
- Tests needing backend-specific SQL (`RunSQL`, FK constraints) carry the SQL as additional `#[case]` parameters.
- `test_atomic_multi_operation_rollback` asserts the **observable** post-condition (all tables dropped) — identical across backends — and additionally cross-checks `DatabaseType::supports_transactional_ddl()` so the divergent execution path is not silently forgotten.
- `test_rollback_fail_with_foreign_key_reference` asserts divergent outcomes explicitly: Postgres / MySQL must `Err`; SQLite is allowed to either `Err` or drop the parent (because SQLite only blocks FK violations on writes, not on `DROP TABLE`). **No `#[ignore]` is used anywhere.**

## Checklist

- [x] Style guidelines followed (tab indent, English-only comments)
- [x] No new clippy warnings in the modified file
- [x] Commit message follows Conventional Commits
- [x] Tests follow AAA pattern with `// Arrange` / `// Act` / `// Assert` comments
- [x] Uses `reinhardt-test` fixtures (no duplicated infrastructure setup)
- [x] Uses `DatabaseConnection` accessors (no raw SeaQuery)
- [x] All discovered failures tracked in follow-up bug issues

## Labels to Apply

- testing
- enhancement

## Related Issues

Fixes #4557
Refs #4582, #4583, #4584, #4585 (follow-up bugs discovered by this PR)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Tests**
  * Expanded migration rollback integration tests to run across PostgreSQL, MySQL, and SQLite.
  * Added a backend-aware test harness to verify rollback outcomes for CREATE TABLE, ADD/ALTER COLUMN, index create/drop, data-preserving rollbacks, and RunSQL reversals.
  * Covers atomic vs non-atomic migrations, foreign-key behaviors, empty or dependent migrations, circular dependency cases, and error scenarios.

[Review Change Stack](https://app.coderabbit.ai/change-stack/kent8192/reinhardt-web/pull/4579?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: tests/integration/tests/migrations/migration_rollback_integration.rs [hunk 1]
```diff
@@ -659,97 +815,109 @@ async fn test_index_rollback(
...
+		.await;
 
+	// Assert: either an explicit error or a no-op skip — both are policy-
+	// compliant, but the system must remain consistent across backends and
+	// must not crash.
 	assert!(
 		rollback_result.is_err() || rollback_result.is_ok(),
-		"Rollback should either fail or skip irreversible operations"
+		"[{backend:?}] rollback should either fail or skip irreversible ops",
 	);
```
[the remark this answers is not in the document]
Comment (kent8192) [reply tests/integration/tests/migrations/migration_rollback_integration.rs:992]: Addressed in 3294d79. Replaced the tautology with an observable-state assertion: the irreversible table must remain after rollback regardless of whether the executor returns `Err` (rejects irreversible ops) or `Ok` (skips them). A successful return must never silently drop user data.

File: tests/integration/tests/migrations/migration_rollback_integration.rs [hunk 2]
```diff
@@ -1069,32 +1285,24 @@ async fn test_circular_dependency_rollback(
...
-	let apply_result = executor.apply_migrations(&[migration_a.clone()]).await;
-
-	// The system should either:
-	// 1. Fail with a circular dependency error
-	// 2. Skip the migration due to missing dependency
-	// 3. Handle it gracefully in some other way
+	// Act: apply (likely errors due to missing/circular dependency)
+	let apply_result = executor
+		.apply_migrations(std::slice::from_ref(&migration_a))
+		.await;
```
[the remark this answers is not in the document]
Comment (kent8192) [reply tests/integration/tests/migrations/migration_rollback_integration.rs:1414]: Addressed in 3294d79. Built an actual cycle by adding `migration_b` that depends on `migration_a` while `migration_a` already depends on `migration_b`, and applied them together so the resolver observes the cycle. The assertion still tolerates either policy (reject with `Err` or break-and-apply) as long as the call returns without panicking.

File: tests/integration/tests/migrations/migration_rollback_integration.rs [hunk 3]
```diff
@@ -306,159 +456,139 @@ async fn test_alter_column_type_rollback(
...
-		reverted_type, "character varying",
-		"Type should be reverted to VARCHAR"
+	// Assert: column still exists with same name; backend-specific type-text
+	// shape varies (e.g. SQLite stores declared types as opaque strings), so
+	// we only assert structural reversibility here. Postgres-specific type-
+	// reversal coverage stays in the dedicated Postgres dialect tests.
+	assert!(
+		column_exists(&connection, "products", "name").await,
+		"[{backend:?}] column should still exist after rollback"
 	);
```
[the remark this answers is not in the document]
Comment (kent8192) [reply tests/integration/tests/migrations/migration_rollback_integration.rs:567]: Addressed in 3294d79. Added a backend-conditional `column_data_type` helper and restored type-reversion assertions for Postgres and MySQL via `information_schema.columns` (post-ALTER is `text`, post-rollback reverts to `character varying` / `varchar`). SQLite is skipped because `information_schema` is unavailable and declared types are stored as opaque strings in `sqlite_master`.

File: tests/integration/tests/migrations/migration_rollback_integration.rs [hunk 4]
```diff
@@ -99,66 +150,171 @@ fn create_auto_pk_column(name: &str, type_def: FieldType) -> ColumnDefinition {
...
+			.fetch_one(&pool)
+			.await
+			.unwrap_or(false)
+		}
+		DatabaseType::Mysql => {
+			let pool = connection.into_mysql().expect("mysql pool unavailable");
+			let count: i64 = sqlx::query_scalar(
+				"SELECT COUNT(*) FROM information_schema.statistics \
+				 WHERE table_schema = DATABASE() AND index_name = ?",
+			)
```
[the remark this answers is not in the document]
Comment (kent8192) [reply tests/integration/tests/migrations/migration_rollback_integration.rs:283]: Addressed in 3294d79. Added a `table_name` parameter to `index_exists` and updated all three backend queries to scope by table: Postgres uses `pg_indexes.tablename`, MySQL adds `table_name = ?` to the `information_schema.statistics` filter, and SQLite uses `sqlite_master.tbl_name`. The two existing call sites pass `"users"` explicitly.

File: tests/integration/tests/migrations/migration_rollback_integration.rs [hunk 5]
```diff
@@ -659,97 +815,109 @@ async fn test_index_rollback(
...
-	// Note: Actual behavior depends on implementation policy
 }
 
-/// Test partial rollback with atomic=false
+/// Test partial rollback with atomic=false.
 ///
-/// **Test Intent**: Verify that atomic=false allows partial rollback
+/// **Test Intent**: Verify that `atomic = false` allows non-atomic rollback.
+/// MySQL is the canonical user here (non-transactional DDL), but Postgres /
+/// SQLite must also succeed.
```
[the remark this answers is not in the document]
Comment (kent8192) [reply tests/integration/tests/migrations/migration_rollback_integration.rs]: Addressed in 3294d79. Rewrote the docstring to accurately describe what the test does: a single-operation `atomic = false` migration applied and rolled back on every backend. Multi-operation partial-failure scenarios under `atomic = false` are intentionally left to the existing error-handling integration suite, which already exercises that path — extending this test would duplicate coverage.

File: tests/integration/tests/migrations/migration_rollback_integration.rs [hunk 6]
```diff
@@ -659,97 +815,109 @@ async fn test_index_rollback(
...
-	// Migration with atomic=false
 	let migration = Migration {
 		app_label: "testapp".to_string(),
 		name: "0001_non_atomic".to_string(),
 		operations: vec![Operation::CreateTable {
-			name: leak_str("table1").to_string(),
+			name: leak_str("non_atomic_table").to_string(),
 			columns: vec![create_auto_pk_column("id", FieldType::Integer)],
 			constraints: vec![],
 			without_rowid: None,
```
[the remark this answers is not in the document]
Comment (kent8192) [reply tests/integration/tests/migrations/migration_rollback_integration.rs:1024]: Addressed in 3294d79 (same docstring rewrite as the line-905 thread). The rewritten docstring no longer claims multi-operation partial-rollback semantics; instead it documents the single-operation non-atomic path that the test actually exercises, with a pointer to the error-handling integration suite for the multi-op partial-failure scenarios.
